From 9b245791da35efed7d3745ca5ca35d6cc73b49de Mon Sep 17 00:00:00 2001
From: Rudi <r.ihle@s-t.de>
Date: Thu, 24 Dec 2015 08:44:50 +0100
Subject: [PATCH] video: mxc_hdmi: Consider modes that differ only in aspect
 ratio as duplicate

Certain CEA modes differ from each other only by the aspect ratio. If an EDID
indicates support for both modes of such an otherwise identical pair, we create
a mode list entry only for the first one. This is done because our sysfs cannot
distinguish them.

Signed-off-by: Rudi <r.ihle@s-t.de>
---
 drivers/video/mxc/mxc_hdmi.c | 15 ++++++++++++++-
 1 file changed, 14 insertions(+), 1 deletion(-)

diff --git a/drivers/video/mxc/mxc_hdmi.c b/drivers/video/mxc/mxc_hdmi.c
index f71c1dd..182ae2a 100644
--- a/drivers/video/mxc/mxc_hdmi.c
+++ b/drivers/video/mxc/mxc_hdmi.c
@@ -1949,7 +1949,7 @@ static void mxc_hdmi_notify_fb(struct mxc_hdmi *hdmi, bool force_all)
 static void mxc_hdmi_create_modelist(struct mxc_hdmi *hdmi, int from_edid)
 {
 	struct fb_videomode mode;
-	int i, vic, mode_count = ARRAY_SIZE(mxc_cea_mode);
+	int i, j, skip, vic, mode_count = ARRAY_SIZE(mxc_cea_mode);
 	const struct fb_videomode *mode_data = mxc_cea_mode;
 
 	if (from_edid) {
@@ -1975,6 +1975,19 @@ static void mxc_hdmi_create_modelist(struct mxc_hdmi *hdmi, int from_edid)
 
 		mode = mode_data[i];
 
+		/* for dual aspect ratio modes, insert the first one only */
+		if (mode.vmode & FB_VMODE_ASPECT_MASK) {
+			skip = 0;
+			for (j = i - 1; !skip && j >= 0; j--) {
+				skip = mode.xres == mode_data[j].xres &&
+				       mode.yres == mode_data[j].yres &&
+				       mode.refresh == mode_data[j].refresh &&
+				       (mode.vmode ^ FB_VMODE_ASPECT_MASK) == mode_data[j].vmode;
+			}
+			if (skip)
+				continue;
+		}
+
 		/* TODO: Discuss if we should always set default modes as standard */
 		if (!from_edid /*&& ignore_edid*/)
 			mode.flag |= FB_MODE_IS_STANDARD;
-- 
1.9.1

